sklearn/manifold/_isomap.py

```diff
@@ -186,9 +190,46 @@ def _fit_transform(self, X):
             n_jobs=self.n_jobs,
         )
 
-        self.dist_matrix_ = graph_shortest_path(
-            kng, method=self.path_method, directed=False
-        )
+        # Compute the number of connected components, and connect the different
+        # components to be able to compute a shortest path between all pairs
+        # of samples in the graph.
+        n_connected_components, labels = connected_components(kng)
+        if n_connected_components > 1:
+            if self.metric == "precomputed":
+                raise RuntimeError(
+                    "The number of connected components of the neighbors graph"
+                    f" is {n_connected_components} > 1. The graph cannot be "
+                    "completed with metric='precomputed', and Isomap cannot be"
+                    "fitted. Increase the number of neighbors to avoid this "
+                    "issue."
+                )
+            warnings.warn(
+                "The number of connected components of the neighbors graph "
+                f"is {n_connected_components} > 1. Completing the graph to fit"
+                " Isomap might be slow. Increase the number of neighbors to "
+                "avoid this issue.",
+                stacklevel=2,
+            )
+
+            for i in range(n_connected_components):
+                idx_i = np.where(labels == i)[0]
+                Xi = X[idx_i]
```

Comment:
I don't recall if we can still bypass the `_validate_data` by using `_safe_indexing` here? Since `NotAnArray` cannot be indexed I am almost sure that it will not be enough.

Reply:
We can use `self.nbrs_._fit_X`, which has been validated by `NearestNeighbors`.

Reply:
Good point, it would avoid a copy most probably.